rainier-core/src/main/scala/com/stripe/rainier/core/Generator.scala

```diff
@@ -6,33 +6,53 @@ import com.stripe.rainier.sampler.RNG
 /**
   * Generator trait, for posterior predictive distributions to be forwards sampled during sampling
   */
-trait Generator[T] { self =>
+sealed trait Generator[T] { self =>
+  import Generator.{Const, From}
+
   def requirements: Set[Real]
 
   def get(implicit r: RNG, n: Numeric[Real]): T
 
-  def map[U](fn: T => U): Generator[U] = new Generator[U] {
-    val requirements: Set[Real] = self.requirements
-    def get(implicit r: RNG, n: Numeric[Real]): U = fn(self.get)
+  def map[U](fn: T => U): Generator[U] = self match {
+    case Const(reqs, t)     => Const(reqs, fn(t))
+    case From(reqs, fromFn) => From(reqs, (r, n) => fn(fromFn(r, n)))
   }
 
-  def flatMap[U](fn: T => Generator[U]): Generator[U] = new Generator[U] {
-    val requirements: Set[Real] = self.requirements
-    def get(implicit r: RNG, n: Numeric[Real]): U = fn(self.get).get
+  def flatMap[U](fn: T => Generator[U]): Generator[U] = self match {
+    case Const(reqsL, t) =>
+      fn(t) match {
+        case Const(reqsR, u)     => Const(reqsL ++ reqsR, u)
+        case From(reqsR, fromFn) => From(reqsL ++ reqsR, fromFn)
+      }
+    case From(reqs, fromFn) =>
+      From(reqs, { (r, n) =>
+        fn(fromFn(r, n)) match {
+          case Const(_, u)          => u
+          case From(_, innerFromFn) => innerFromFn(r, n)
+        }
+      })
   }
 
-  def zip[U](other: Generator[U]): Generator[(T, U)] = new Generator[(T, U)] {
-    val requirements: Set[Real] = self.requirements ++ other.requirements
-    def get(implicit r: RNG, n: Numeric[Real]): (T, U) = (self.get, other.get)
+  def zip[U](other: Generator[U]): Generator[(T, U)] = {
+    val reqs: Set[Real] = self.requirements ++ other.requirements
+    (self, other) match {
+      case (Const(_, t), Const(_, u)) => Const(reqs, (t, u))
+      case (From(_, lf), From(_, rf)) =>
+        From(reqs, (r, n) => (lf(r, n), rf(r, n)))
+      case (Const(_, t), From(_, rf)) => From(reqs, (r, n) => (t, rf(r, n)))
+      case (From(_, lf), Const(_, u)) => From(reqs, (r, n) => (lf(r, n), u))
+    }
   }
 
-  def repeat(k: Real): Generator[Seq[T]] = new Generator[Seq[T]] {
-    val requirements: Set[Real] = self.requirements
-    def get(implicit r: RNG, n: Numeric[Real]): Seq[T] =
-      0.until(n.toInt(k)).map { _ =>
-        self.get
-      }
-  }
+  def repeat(k: Real): Generator[Seq[T]] =
+    Generator.require(requirements)(self match {
```

Comment:
I can use `From[Seq[T]](requirements)(....)` directly here instead of `Generator.require` if you like, since we're still inside the trait. It's up to you.

Reply:
This way seems fine, thanks.